src/arch/src/aarch64/gic/regs/mod.rs

```diff
@@ -167,27 +167,43 @@ impl MmioReg for SimpleReg {
 #[derive(Debug, Default, Versionize)]
 pub struct GicState {
     dist: Vec<GicRegData<u32>>,
-    rdist: Vec<Vec<GicRegData<u32>>>,
-    icc: Vec<icc_regs::VgicSysRegsState>,
+    vcpu_states: Vec<GicVcpuState>,
+}
+
+/// Structure used for serializing the state of the GIC registers for a specific vCPU
+#[derive(Debug, Default, Versionize)]
+pub struct GicVcpuState {
+    rdist: Vec<GicRegData<u32>>,
+    icc: icc_regs::VgicSysRegsState,
 }
 
 /// Save the state of the GIC device.
 pub fn save_state(fd: &DeviceFd, mpidrs: &[u64]) -> Result<GicState> {
     // Flush redistributors pending tables to guest RAM.
     super::save_pending_tables(fd)?;
 
+    let mut vcpu_states = Vec::with_capacity(mpidrs.len());
+    for mpidr in mpidrs {
+        vcpu_states.push(GicVcpuState {
+            rdist: redist_regs::get_redist_regs(fd, *mpidr)?,
+            icc: icc_regs::get_icc_regs(fd, *mpidr)?,
+        })
+    }
+
     Ok(GicState {
         dist: dist_regs::get_dist_regs(fd)?,
-        rdist: redist_regs::get_redist_regs(fd, mpidrs)?,
-        icc: icc_regs::get_icc_regs(fd, mpidrs)?,
+        vcpu_states,
     })
 }
 
 /// Restore the state of the GIC device.
 pub fn restore_state(fd: &DeviceFd, mpidrs: &[u64], state: &GicState) -> Result<()> {
     dist_regs::set_dist_regs(fd, &state.dist)?;
-    redist_regs::set_redist_regs(fd, mpidrs, &state.rdist)?;
-    icc_regs::set_icc_regs(fd, mpidrs, &state.icc)?;
+
+    for (mpidr, vcpu_state) in mpidrs.iter().zip(&state.vcpu_states) {
```

Comment:
I would add sanity checks for all places where we do these `zip()`s.

Let's fail fast and explicit if the `mpidrs.len()` doesn't match the `vcpu_states.len()`.

Reply:
Done